### DISCUSSION TITLE
Issue with using the JMX reporter with platform MBean server

### DISCUSSION THREAD

**Discussion by rohit2395** [dropwizard/metrics#3120] [Q&A]
Hi. I am using the latest `metrics-jmx (4.2.12)`  library. 

Here is a snippet of my code

Initialize the metrics registry:
```
MetricRegistry registry = new MetricRegistry();
final JmxReporter reporter = JmxReporter.forRegistry(registry).build();
reporter.start();
```

Register metrics:
```
String metricKey = "metric.key";

// metrics map
HashMap<String,Metric> metricMap = new HashMap<String,Metric>(6);

..
..
registry.register(metricKey,metricMap.get(metricKey));
..
..
```

Get the count using `ManagementFactory.getPlatformMBeanServer()`
```
MBeanServerConnection server = ManagementFactory.getPlatformMBeanServer()
String metricKey = "metric.key";
ObjectName metric = new ObjectName("metrics:name=" + metricKey);
 if (server.isRegistered(metric)){
      Object object = server.getAttribute(metric, "Count");
      result = (Long) object;
}else {
      System.out.println("Metric is not registered : "+ metric);
}
```

The issue is it says that the metric is not registered i.e `server.isRegistered(metric)` returns `false`. 

This works with `metrics-jmx (4.0.7)` but not `metrics-jmx (4.1.0)`

Can someone help me if I need to do something else to achieve this? or is it an issue in the library?

**Comment by the-thing** [comment]
When registering a new metric with `JmxReporter` the name is generated via `com.codahale.metrics.jmx.ObjectNameFactory` which might be slightly different between versions.

```
server.isRegistered(new ObjectName("metrics:name=" + metricKey)) // returns false
server.isRegistered(new ObjectName("metrics:name=" + metricKey + ",type=counters")) // returns true when registered metric is a counter
```


com.codahale.metrics.jmx.ObjectNameFactory is package private in `JmxReporter`, but can provide custom one with a domain name via `com.codahale.metrics.jmx.JmxReporter.Builder#Builder`.

Unfortunately you also need to know the metric types which are hardcoded in the implementation (see usages of `com.codahale.metrics.jmx.JmxReporter.JmxListener#createName`).

```
String domain = "metrics";
ObjectNameFactory objectNameFactory = new DefaultObjectNameFactory();

MetricRegistry registry = new MetricRegistry();
JmxReporter reporter = JmxReporter.forRegistry(registry).createsObjectNamesWith(objectNameFactory).inDomain(domain).build();
reporter.start();

String metricKey = "metric.key";
registry.register(metricKey, new Counter());

MBeanServer server = ManagementFactory.getPlatformMBeanServer();
System.out.println(server.isRegistered(objectNameFactory.createName("counters", domain, metricKey)));
```
